sos/policies/auth/deviceauth.py

```diff
@@ -0,0 +1,307 @@
+        authenticated or an error occurs.
+
+        """
+        token_data = {'grant_type': GRANT_TYPE_DEVICE_CODE,
+                      'client_id': DEVICE_AUTH_CLIENT_ID,
+                      'device_code': self.__device_code}
+
+        if not REQUESTS_LOADED:
+            raise Exception("python3-requests is not installed and is required"
+                            " for obtaining device auth token.")
+        while self._access_token is None:
+            time.sleep(self._interval)
+            try:
+                check_auth_completion = requests.post(self.token_endpoint,
+                                                      data=token_data)
+
+                status_code = check_auth_completion.status_code
+
+                if status_code == 200:
+                    logger.info("The SSO authentication is successful")
+                    self._set_token_data(check_auth_completion.json())
+                if status_code not in [200, 400]:
+                    raise Exception(status_code, check_auth_completion.text)
+                if status_code == 400 and \
+                    check_auth_completion.json()['error'] not in \
+                        ("authorization_pending", "slow_down"):
+                    raise Exception(status_code, check_auth_completion.text)
+            except Exception as e:
+                raise e
+
+    def _set_token_data(self, token_data):
+        """
+        Set the class attributes as per the input token_data received and
+         persist it in the local keyring to avoid
+        visting the browser frequently.
+        :param token_data: Token data containing access_token, refresh_token
+         and their expiry etc.
+        """
+        self._access_token = token_data.get("access_token")
+        self._access_expires_at = datetime.utcnow() + \
+            timedelta(seconds=token_data.get("expires_in"))
+        self._refresh_token = token_data.get("refresh_token")
+        self._refresh_expires_in = token_data.get("refresh_expires_in")
+        if self._refresh_expires_in == 0:
+            self._refresh_expires_at = datetime.max
+        else:
+            self._refresh_expires_at = datetime.utcnow() + \
+                timedelta(seconds=self._refresh_expires_in)
+        self.persist_refresh_token()
+
+    def get_access_token(self):
+        """
+        Get the valid access_token at any given time.
+        :return: Access_token
+        :rtype: string
+        """
+        if self.is_access_token_valid():
+            return self._access_token
+
+        if self.grant_type == "client_credentials":
+            self._use_client_credentials_grant()
+            return self._access_token
+
+        elif self.grant_type == "device_auth":
+            if self.is_refresh_token_valid():
+                self._use_refresh_token_grant()
+                return self._access_token
+            else:
+                return self.request_new_device_code()
```

Comment:
It was an attempt to have both device authorization and password based authorization. But since for now we only have device authorization, I'm removing the grant_type references. We can revisit adding them in the future if we add more auth types.

Reply:
This is still valid (why the CI test stopped detecting it..?).

When we get to this line, `self.request)new_device_code()` is called. That method is just a "proxy" to `self._use_device_code_grant()` as it does executes just that command. And does not return anything.

I suggest:

- removing `request_new_device_code` method and replace its both calls by calling `self._use_device_code_grant()` directly.
- have here two lines:

```
    self._use_device_code_grant()
    return self._access_token
```

Reply:
Ack! Fixed in the next version.